### PR TITLE
introduce `llnl.util.compat` to centralize python version-specific imports

### PR DESCRIPTION
**Pull request by cosmicexplorer**
### Problem

Several places across the repo do `if sys.version_info < ...` checks to determine what to import. This can be error-prone and hard to follow.

Additionally, @trws suggested looking into the use of `six.raise_from()` where possible to preserve exception traceback.

### Solution
- Create a file `lib/spack/spack/util/compat.py` to centralize `if sys.version_info ...` checks across the spack codebase.
- Create another file `.../py2.py` to centralize logic relating to py2/3 compat that is not simply `if sys.version_info` imports.
- Use `six.raise_from()` in several places where it could preserve additional chained exception traceback in python 3.

### Result
This change reduces a boilerplate `if sys.version_info....:` check into a single clean import statement.

### CODE REVIEW

**Comment by adamjstewart** [comment]
This is also how many of the libraries we vendor handle version compatibility nowadays, so I agree that this is the correct approach.

**Comment by cosmicexplorer** [comment]
This is pretty incredible. I seem to have deleted the `compat.py` file. I am going to replace it and add `typing_extensions` to this, since that is currently causing #21538 to fail.

**Comment by cosmicexplorer** [comment]
In addition to the `typing` changes suggested by @trws here (https://github.com/spack/spack/pull/22235#discussion_r593980704), I think we should consider using something from `six` or hacking something ourselves to use the python 3 `raise Exception(...) from e` method of chaining stacktraces.

**Comment by trws** [comment]
Yes, the `six.raise_from` would be really good to use.  We already have six included in the vendor stuff, so it should be usable as is.  Not sure we can really require it, but if you run into a place where that would make sense definitely do it.  No reason not to get the better exceptions.

**Comment by cosmicexplorer** [comment]
@trws @adamjstewart the codecov "patch" test seems to be failing spuriously, but otherwise this is ready for review!

File: .flake8 [hunk 1]
```diff
@@ -35,8 +35,13 @@
...
 # - W503: see above, already ignored for line-breaks
 #
+# Import failure ignores: these are handled by isort.
+# - I100: Incorrect import order.
+# - I201: Missing newline between import groups.
+# - I202: Additional newline within an import group.
+#
 [flake8]
-ignore = E129,E221,E241,E272,E731,W503,W504,F999,N801,N813,N814,F403,F405
+ignore = E129,E221,E241,E272,E731,W503,W504,F999,N801,N813,N814,F403,F405,I100,I201,I202
```
Comment (adamjstewart) [inline .flake8]: Spack no longer uses `flake8-import-order`, so these changes can be removed

File: .isort.cfg [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[settings]
+profile=black
+known_first_party=spack
+known_archspec=archspec
+known_ruamel=ruamel
+known_compat=compat
+known_llnl=llnl
+known_third_party=six
+known_typing=typing,typing_extensions
+sections=FUTURE,STDLIB,TYPING,THIRDPARTY,COMPAT,ARCHSPEC,RUAMEL,LLNL,FIRSTPARTY,LOCALFOLDER
```
Comment (adamjstewart) [inline .isort.cfg]: Our isort configuration is now stored in `pyproject.toml`, so this can be removed.

File: pyproject.toml [hunk 3]
```diff
@@ -1,21 +1,28 @@
...
+    "ARCHSPEC", "RUAMEL", "LLNL", "FIRSTPARTY",
+    "LOCALFOLDER",
+  ]
 known_first_party = "spack"
 known_archspec = "archspec"
+known_ruamel = "ruamel"
+known_compat = "compat"
 known_llnl = "llnl"
+known_third_party = "six"
+known_typing = ["typing", "typing_extensions"]
```
Comment (adamjstewart) [inline pyproject.toml]: These libraries are all third-party libraries, and don't need an additional import section. Why are these changes necessary?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
There are a bunch of changes unrelated to `compat` in this PR that should probably be removed.

File: .flake8 (hunk 1 above)
Comment (cosmicexplorer) [reply .flake8]: Thanks! Done.

File: .isort.cfg (hunk 2 above)
Comment (cosmicexplorer) [reply .isort.cfg]: Thanks! Done.

File: pyproject.toml (hunk 3 above)
Comment (cosmicexplorer) [reply pyproject.toml]: Thanks so much!! I have removed these individual import sections and rolled them all into `"THIRDPARTY"`.

**Comment by cosmicexplorer** [comment]
Codecov is failing I suspect because of the `six.raise_from()` calls which re-raise exceptions that aren't adequately covered by tests. I believe this can be considered to be passing.

**Review by tgamblin — CHANGES_REQUESTED** [review]
Can you remove changes to everything in external and pull `compat.py` into `llnl.util`?  We don't want to bother maintaining changes to external packages -- the stuff in `lib/spack/external` should be (almost entirely) straight from the original source.  This makes it easier to see what changed when we update vendored dependencies too.

**Comment by cosmicexplorer** [comment]
@tgamblin: yes, done! That will almost definitely fix the codecov issues I was having too. Will ping again to review after making sure I haven't broken anything.

**Comment by cosmicexplorer** [comment]
Ready for review!! Thanks again.

File: lib/spack/spack/mirror.py [hunk 4]
```diff
@@ -69,6 +65,10 @@ def __init__(self, fetch_url, push_url=None,
         self._push_url = push_url
         self._name = name
 
+    def __eq__(self, other):
+        return (self._fetch_url == other._fetch_url and
+                self._push_url == other._push_url)
+
```
Comment (tgamblin) [inline lib/spack/spack/mirror.py]: why's this needed?

File: lib/spack/spack/mirror.py [hunk 5]
```diff
@@ -238,6 +246,9 @@ def __init__(self, mirrors=None, scope=None):
                 mirrors.items() if mirrors is not None else
                 spack.config.get('mirrors', scope=scope).items()))
 
+    def __eq__(self, other):
+        return self._mirrors == other._mirrors
+
```
Comment (tgamblin) [inline lib/spack/spack/mirror.py]: same question here

File: lib/spack/spack/mirror.py [hunk 6]
```diff
@@ -81,12 +81,20 @@ def from_yaml(stream, name=None):
...
 
     @staticmethod
     def from_json(stream, name=None):
-        d = sjson.load(stream)
-        return Mirror.from_dict(d, name)
+        try:
+            d = sjson.load(stream)
+            return Mirror.from_dict(d, name)
+        except Exception as e:
+            tty.debug(e)
```
Comment (tgamblin) [inline lib/spack/spack/mirror.py]: leftover debug print?

File: lib/spack/spack/mirror.py [hunk 7]
```diff
@@ -251,12 +262,20 @@ def from_yaml(stream, name=None):
...
 
     @staticmethod
     def from_json(stream, name=None):
-        d = sjson.load(stream)
-        return MirrorCollection(d)
+        try:
+            d = sjson.load(stream)
+            return MirrorCollection(d)
+        except Exception as e:
+            tty.debug(e)
```
Comment (tgamblin) [inline lib/spack/spack/mirror.py]: we shouldn't need these as there is already debug handling for printing out the exception trace in `spack.error`

File: lib/spack/spack/spec.py [hunk 8]
```diff
@@ -2183,7 +2177,9 @@ def from_yaml(stream):
             data = yaml.load(stream)
             return Spec.from_dict(data)
         except yaml.error.MarkedYAMLError as e:
-            raise syaml.SpackYAMLError("error parsing YAML spec:", str(e))
+            raise six.raise_from(
+                syaml.SpackYAMLError("error parsing YAML spec:", str(e)),
+                e)
```
Comment (tgamblin) [inline lib/spack/spack/spec.py]: for lines like this you should probably just put the paren on the next line to match black formatting (which we'll eventually use)

File: lib/spack/spack/test/mirror.py [hunk 9]
```diff
@@ -149,6 +151,100 @@ def test_all_mirror(
...
+    ]
+)
+def test_invalid_json_mirror_collection(invalid_json, error_message):
+    with pytest.raises(sjson.SpackJSONError) as e:
+        spack.mirror.MirrorCollection.from_json(invalid_json)
+    exc_msg = str(e.value)
+    assert exc_msg.startswith("error parsing JSON mirror collection:")
+    assert error_message in exc_msg
+
+
```
Comment (tgamblin) [inline lib/spack/spack/test/mirror.py]: More tests are good -- but why are they in this PR?

File: lib/spack/spack/cmd/style.py [hunk 10]
```diff
@@ -295,12 +295,19 @@ def run_mypy(mypy_cmd, file_list, args):
         "--package", "spack",
         "--package", "llnl",
         "--show-error-codes",
+        "--follow-imports=silent",
+        "--ignore-missing-imports",
```
Comment (becker33) [inline lib/spack/spack/cmd/style.py]: This scares me -- can you explain why it's safe?

File: lib/spack/spack/mirror.py [hunk 11]
```diff
@@ -69,6 +65,10 @@ def __init__(self, fetch_url, push_url=None,
         self._push_url = push_url
         self._name = name
 
+    def __eq__(self, other):
+        return (self._fetch_url == other._fetch_url and
+                self._push_url == other._push_url)
```
Comment (becker33) [inline lib/spack/spack/mirror.py]: This seems orthogonal to the PR in  general

File: lib/spack/spack/mirror.py [hunk 12]
```diff
@@ -238,6 +246,9 @@ def __init__(self, mirrors=None, scope=None):
                 mirrors.items() if mirrors is not None else
                 spack.config.get('mirrors', scope=scope).items()))
 
+    def __eq__(self, other):
+        return self._mirrors == other._mirrors
```
Comment (becker33) [inline lib/spack/spack/mirror.py]: Also seems like a separate PR

File: lib/spack/spack/test/mirror.py [hunk 13]
```diff
@@ -149,6 +151,100 @@ def test_all_mirror(
...
+    [
+        spack.mirror.Mirror(
+            'https://example.com/fetch',
+            'https://example.com/push',
+        ),
+    ],
+)
+def test_serde_mirror(mirror):
+    mirror_yaml = mirror.to_yaml()
+    assert spack.mirror.Mirror.from_yaml(mirror_yaml) == mirror
```
Comment (becker33) [inline lib/spack/spack/test/mirror.py]: Are the `__eq__` methods for mirrors in the PR just to make this test possible?

File: lib/spack/spack/test/mirror.py [hunk 14]
```diff
@@ -149,6 +151,100 @@ def test_all_mirror(
...
+@pytest.mark.parametrize(
+    "mirror",
+    [
+        spack.mirror.Mirror(
+            'https://example.com/fetch',
+            'https://example.com/push',
+        ),
+    ],
+)
+def test_serde_mirror(mirror):
```
Comment (becker33) [inline lib/spack/spack/test/mirror.py]: I think this should be `serdes`, but I could definitely be wrong. I think `serde` is the Rust framework for `serdes`.

In other test modules, the serdes methods are named `*_roundtrip_[lang]` so we haven't had to think about it before.

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/mirror.py (hunk 9 above)
Comment (cosmicexplorer) [reply lib/spack/spack/test/mirror.py]: To pass codecov. I'll try removing them now that I've reverted everything in `lib/spack/external`.

File: lib/spack/spack/mirror.py (hunk 4 above)
Comment (cosmicexplorer) [reply lib/spack/spack/mirror.py]: In order to add testing so we could pass codecov. I have reverted it for now and will see if that passes.

File: lib/spack/spack/test/mirror.py (hunk 13 above)
Comment (cosmicexplorer) [reply lib/spack/spack/test/mirror.py]: That's correct, which was in order to pass codecov. I have reverted these and am going to see whether it passes codecov anyway after I took Todd's suggestion to revert all the changes in `lib/spack/external`.

File: lib/spack/spack/test/mirror.py (hunk 14 above)
Comment (cosmicexplorer) [reply lib/spack/spack/test/mirror.py]: Yes, `serde` is the rust framework for de/serialization based on the colloquial term! I have removed this testing as it seems unrelated to the thrust of the PR.

File: lib/spack/spack/mirror.py (hunk 6 above)
Comment (cosmicexplorer) [reply lib/spack/spack/mirror.py]: No, it was stolen from `Spec.from_json()`, where I assumed the debug print was there for a reason. I have reverted this and removed `tty.debug(e)` from the `Spec` method too.

File: lib/spack/spack/cmd/style.py (hunk 10 above)
Comment (cosmicexplorer) [reply lib/spack/spack/cmd/style.py]: It turns out to be unnecessary, and I have removed it. I believe it may have been necessary before we made other changes to mypy in this repo several months ago.

File: lib/spack/spack/mirror.py (hunk 11 above)
Comment (cosmicexplorer) [reply lib/spack/spack/mirror.py]: Yes, and it has been reverted!

File: lib/spack/spack/mirror.py (hunk 12 above)
Comment (cosmicexplorer) [reply lib/spack/spack/mirror.py]: Yes, and this has been reverted!

File: lib/spack/spack/mirror.py (hunk 5 above)
Comment (cosmicexplorer) [reply lib/spack/spack/mirror.py]: In order to add testing so we could pass codecov. I have reverted it for now and will see if that passes.

File: lib/spack/spack/mirror.py (hunk 7 above)
Comment (cosmicexplorer) [reply lib/spack/spack/mirror.py]: See https://github.com/spack/spack/pull/21720#discussion_r789192991.

File: lib/spack/spack/spec.py (hunk 8 above)
Comment (cosmicexplorer) [reply lib/spack/spack/spec.py]: Done!

**Comment by cosmicexplorer** [comment]
Ok, so codecov is still failing, so I'm going to unrevert the patches that add more test coverage and see if it passes.

**Comment by cosmicexplorer** [comment]
Actually, I'm going to split the `six.raise_from()` changes into another PR to avoid triggering codecov here so we can focus on the `llnl.util.compat` changes!

File: lib/spack/llnl/util/compat.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+# crashes of the subprocess. Figure out how to make 'spawn' work.
+#
+# See:
+# * https://github.com/spack/spack/pull/18124
+# * https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
+# * https://bugs.python.org/issue33725
+if sys.version_info >= (3, 8):
+    fork_context = multiprocessing.get_context('fork')
+else:
+    fork_context = multiprocessing
```
Comment (adamjstewart) [inline lib/spack/llnl/util/compat.py]: I didn't think we were still doing this. Didn't we properly define pickling procedures for all objects that needed it?

Comment (cosmicexplorer) [reply lib/spack/llnl/util/compat.py]: Yes, it looks like this is no longer used anywhere. I will remove it!

**Comment by cosmicexplorer** [comment]
Ok codecov is going absolutely bonkers and claiming lines I didn't change would affect the coverage. I was able to fix this last time by amending the commit and force-pushing twice in a row -- will try to do the same here. Otherwise, tests are passing.

**Comment by cosmicexplorer** [comment]
Ok, codecov magically went way up after doing that, but it looks like it's not quite passing yet. I'll figure out what it doesn't like and make that pass.

**Comment by cosmicexplorer** [comment]
Oh! It's because I added a new method which wasn't used anywhere for some reason. I reverted that and I think it should pass now.

**Review by becker33 — DISMISSED** [review]

**Comment by cosmicexplorer** [comment]
@becker33 had to push again to fix the spurious codecov error, I suspect this time should work. Didn't make any changes.

**Review by becker33 — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]
